### PR TITLE
unix,win: add uv_spawn option to set child CPU affinity mask

### PR DESCRIPTION
**Pull request by bradking**
Implement it on Linux, FreeBSD, and Windows for now, and fail with `UV_ENOTSUP` on other platforms. (EDIT: Added FreeBSD)

Issue: #1389

### CODE REVIEW

File: src/unix/core.c [hunk 1]
```diff
@@ -41,6 +41,10 @@
 #include <sys/resource.h> /* getrusage */
 #include <pwd.h>
 
+#if defined(__linux__)
+# include <sched.h>
```
Comment (krytarowski) [inline src/unix/core.c]: This header is available on more OSes, at least on NetBSD. Is there need to restrict it to Linux-only?

Comment (bradking) [reply src/unix/core.c]: Not in the long-term, but for now my goal is to get the API worked out with sample implementaions on Linux and Windows.  Additional ports may follow.

Comment (krytarowski) [reply src/unix/core.c]: Please enable it on all OSes unconditionally. (unless there exists a UNIX-like system without this header)

Comment (bradking) [reply src/unix/core.c]: Done!

File: test/run-tests.c [hunk 2]
```diff
@@ -195,5 +195,45 @@ static int maybe_run_test(int argc, char **argv) {
...
+    cpumask_size = uv_cpumask_size();
+    ASSERT(cpumask_size > 0);
+
+    cpu = atoi(argv[2]);
+    ASSERT(cpu >= 0);
+    ASSERT(cpu < cpumask_size);
+
+    /* verify the mask has the cpu we expect */
+#ifdef _WIN32
+    r = GetProcessAffinityMask(GetCurrentProcess(), &procmask, &sysmask)?1:0;
```
Comment (bnoordhuis) [inline test/run-tests.c]: Why the ternary?  Couldn't you just assert that `r != 0`?

File: test/test-spawn.c [hunk 3]
```diff
@@ -1367,6 +1367,79 @@ TEST_IMPL(spawn_setgid_fails) {
...
+#endif
+
+  cpumask_size = uv_cpumask_size();
+  ASSERT(cpumask_size > 0);
+
+  /* find a cpu we can use */
+  cpu = cpumask_size;
+#ifdef _WIN32
+  r = GetProcessAffinityMask(GetCurrentProcess(), &procmask, &sysmask)?1:0;
+  ASSERT(r == 1);
```
Comment (bnoordhuis) [inline test/test-spawn.c]: Same question.

File: test/run-tests.c (hunk 2 above)
Comment (bradking) [reply test/run-tests.c]: Oops, that was left from an earlier iteration of the code.  Updated.

**Comment by bnoordhuis** [comment]
@libuv/collaborators Anyone else want to sign off on this?

**Comment by krytarowski** [comment]
I think that such machine-dependent API should be added along with more OSes at ones, in order to assure that API is reusable for others.

**Comment by krytarowski** [comment]
On NetBSD API is very similar to the Linux one, but uses opaque types; but for start at least all current Tier1 ports should get this functional.

File: docs/src/process.rst [hunk 4]
```diff
@@ -172,6 +174,21 @@ Public members
         This is not supported on Windows, :c:func:`uv_spawn` will fail and set the error
         to ``UV_ENOTSUP``.
 
+.. c:member:: uv_process_options_t.cpumask
+.. c:member:: uv_process_options_t.cpumask_size
+
+    Libuv can set the child process' CPU affinity mask.  This happens when
+    `cpumask` is non-NULL.  It must point at an array of char values
```
Comment (santigimeno) [inline docs/src/process.rst]: s/at/to

File: docs/src/process.rst [hunk 5]
```diff
@@ -32,6 +32,8 @@ Data types
             uv_stdio_container_t* stdio;
             uv_uid_t uid;
             uv_gid_t gid;
+            char*  cpumask;
```
Comment (santigimeno) [inline docs/src/process.rst]: nit: only one space between char* and cpumask

File: include/uv.h [hunk 6]
```diff
@@ -956,6 +956,19 @@ typedef struct uv_process_options_s {
...
+    `cpumask` is non-NULL.  It must point at an array of char values
+    of length `cpumask_size`, whose value must be at least that returned by
+    uv_cpumask_size().  Each byte in the mask can be either zero (false)
+    or non-zero (true) to indicate whether the corresponding processor at
+    that index is included.
+
+    If enabled on an unsupported platform, uv_spawn() will fail with
+    UV_ENOTSUP.
+   */
+  char*  cpumask;
```
Comment (santigimeno) [inline include/uv.h]: only one space

File: docs/src/process.rst [hunk 7]
```diff
@@ -172,6 +174,21 @@ Public members
...
         to ``UV_ENOTSUP``.
 
+.. c:member:: uv_process_options_t.cpumask
+.. c:member:: uv_process_options_t.cpumask_size
+
+    Libuv can set the child process' CPU affinity mask.  This happens when
+    `cpumask` is non-NULL.  It must point at an array of char values
+    of length `cpumask_size`, whose value must be at least that returned by
+    :c:func:`uv_cpumask_size`.  Each byte in the mask can be either
+    zero (false) or non-zero (true) to indicate whether the corresponding
```
Comment (santigimeno) [inline docs/src/process.rst]: Out curiosity: any reason not to do this bitwise instead of byte wise?

File: test/test-spawn.c [hunk 8]
```diff
@@ -1367,6 +1367,79 @@ TEST_IMPL(spawn_setgid_fails) {
...
+  newmask = (char*)calloc(cpumask_size, 1);
+  ASSERT(newmask != NULL);
+  newmask[cpu] = 1;
+  options.cpumask_size = (size_t)cpumask_size;
+  options.cpumask = newmask;
+
+  /* tell the child which one it should get */
+  options.args[2] = cpustr;
+  options.args[3] = "dummy"; /* need 4 args for test/run-tests.c dispatch */
+
```
Comment (santigimeno) [inline test/test-spawn.c]: I woul add a test where cpumask_size were less than required to check it fails with `UV_EINVAL`

File: src/win/process.c [hunk 9]
```diff
@@ -1083,6 +1095,45 @@ int uv_spawn(uv_loop_t* loop,
...
   }
 
+  if (options->cpumask != NULL) {
+    /* The child is currently suspended.  Set its process affinity
+       or terminate it if we can't.  */
+    int i;
+    int cpumasksize;
+    DWORD_PTR sysmask;
+    DWORD_PTR oldmask;
+    DWORD_PTR newmask = 0;
```
Comment (santigimeno) [inline src/win/process.c]: separate declaration from assignment

File: docs/src/misc.rst [hunk 10]
```diff
@@ -223,6 +223,11 @@ API
 
     Frees the `cpu_infos` array previously allocated with :c:func:`uv_cpu_info`.
 
+.. c:function:: int uv_cpumask_size(void)
+
+    Returns the maximum size of the mask used for process/thread affinities,
+    or ``UV_ENOTSUP`` if affinities are not supported on the current platform.
+
```
[the remark this answers is not in the document]
Comment (bradking) [reply docs/src/misc.rst]: I added the `versionadded` label.

File: docs/src/process.rst [hunk 11]
```diff
@@ -172,6 +174,21 @@ Public members
...
+    processor at that index is included.
+
+    .. note::
+
+        If enabled on an unsupported platform, :c:func:`uv_spawn` will fail
+        with ``UV_ENOTSUP``.
+
```
[the remark this answers is not in the document]
Comment (bradking) [reply docs/src/process.rst]: I added the `versionadded` label.

File: test/test-spawn.c (hunk 8 above)
Comment (bradking) [reply test/test-spawn.c]: Good idea.  Added as another test case.

File: docs/src/process.rst (hunk 7 above)
Comment (bradking) [reply docs/src/process.rst]: The API was worked out by discussion in #1389 and based on previous discussion in another MR for thread affinity.  IIUC the main reason for bytewise masking is simplicity.

**Comment by santigimeno** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/526/

File: src/win/process.c [hunk 12]
```diff
@@ -1083,6 +1095,46 @@ int uv_spawn(uv_loop_t* loop,
...
+    }
+
+    if (!SetProcessAffinityMask(info.hProcess, newmask)) {
+      err = GetLastError();
+      TerminateProcess(info.hProcess, 1);
+      goto done;
+    }
+
+    /* The process affinity of the child is set.  Let it run.  */
+    ResumeThread(info.hThread);
```
Comment (santigimeno) [inline src/win/process.c]: Should we check the return value of `ResumeThread()` ?

Comment (bradking) [reply src/win/process.c]: Good catch, thanks.  Updated.

**Comment by santigimeno** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/528/

**Comment by bradking** [comment]
Now with two approvals, can this be merged please?

**Comment by bnoordhuis** [comment]
This can land after master has been synced again, see https://github.com/libuv/libuv/pull/1630.  It's possible this PR might need a rebase after that but I'll ping you if it does.

**Comment by bradking** [comment]
@bnoordhuis with #1753 merged, this MR may now be merged IIUC.  I've rebased on `master` to resolve conflicts.

**Comment by santigimeno** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/733/

**Comment by bradking** [comment]
The CI looks pretty happy.  Two builds had test failures but they look unrelated or spurious.  One build (zos) however [failed to clone the repo](https://ci.nodejs.org/job/libuv-test-commit-zos/491/nodes=zos-s390x/console).

**Comment by santigimeno** [comment]
Landed in https://github.com/libuv/libuv/commit/6521405c07485eb77c625c01b7afc6b8dd6c2e61 and https://github.com/libuv/libuv/commit/20cc969e5db3467bddaa35f48ba051a7ae3b3653. Thanks!

**Pull request closed without merging**